browser/src/layer/marker/TextInput.js

```diff
@@ -289,16 +290,32 @@ L.TextInput = L.Layer.extend({
 	// Marks the content of the textarea/contenteditable as selected,
 	// for system clipboard interaction.
 	select: function() {
-		this._textArea.select();
+		this._setSelectionRange(0, this.getPlainTextContent().length);
 	},
 
 	getValue: function() {
-		var value = this._textArea.value;
+		var value = this.getPlainTextContent();
 		if (this._map.formulabar && this._map.formulabar.hasFocus())
-			value = this._map.formulabar.getValue();
+			value =  ' ' + this._map.formulabar.getValue() + ' ';
 		return value;
 	},
 
+	getPlainTextContent: function() {
+		return 	this._textArea.textContent;
+	},
+
+	getHTML: function() {
+		return 	this._textArea.innerHTML;
+	},
+
+	resetContent: function() {
+		this._textArea.innerHTML = this._initialContent;
+		this._textArea.firstChild.textContent = '\xa0';
```

Comment:
Do we really need this ? surely we have the <img - now which doesn't need a textContent ? =)

Reply:
As I told above `_onInput` expects extra spaces. I'd very like to refactor `_onInput` and get rid of extra spaces, anyway I'd need some enlightenment on part of `_onInput` where we check for content length < 1 and < 2: by reomoving extra spaces it is no more possible to perform these checks:

		// We use a different leading and terminal space character
		// to differentiate backspace from delete, then replace the character.
		if (content.length < 1 || !this._hasPreSpace()) { // missing initial space
			window.app.console.log('Sending backspace');
			if (!ignoreBackspace)
				this._removeTextContent(1, 0);
			this._emptyArea();
			return;
		}
		if (!this._hasPostSpace()) { // missing trailing space.
			window.app.console.log('Sending delete');
			this._removeTextContent(0, 1);
			this._emptyArea();
			return;
		}
		if (content.length < 2) {
			window.app.console.log('Missing terminal nodes: ' + this._deleteHint);
			if (this._deleteHint == 'backspace' ||
				this._getSelectionStart() === 0)
			{
				if (!ignoreBackspace)
					this._removeTextContent(1, 0);
			}
			else if (this._deleteHint == 'delete' ||
				this._getSelectionStart() === 1)
				this._removeTextContent(0, 1);
			else
				window.app.console.log('Cant detect delete or backspace');
			this._emptyArea();
			return;
		}

For the former check the original if clause was:

    if (content.length < 1 || content[0] !== preSpaceChar)

    
Maybe they are no more needed now but on first pass I preferred to be conservative, since I didn't want to miss some use cases.

Reply:
Well - lets clean this up to make it maintainable - if we don't believe that is used anymore - lets get rid of it. One of the problems in the past was that auto-correction and/or random Android input methods would not only move your cursor to 'help' but also change the content of the text area. I think that's less likely with our content-editable approach and HTML - it'd be too risky / complex for them =)

So lets drop this content.length < 1 complexity - I'm sure we'll get new complexity back for something else =)

Reply:
TextInput simplified; extra spaces and text nodes inside <img> delimiters removed.